src/wallet/wallet.cpp

```diff
@@ -2861,26 +2861,36 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
     }
 
     //if we're doing only denominated, we need to round up to the nearest smallest denomination
-    if(nCoinType == ONLY_DENOMINATED) {
+    if (nCoinType == ONLY_DENOMINATED) {
         std::vector<CAmount> vecPrivateSendDenominations = CPrivateSend::GetStandardDenominations();
-        CAmount nSmallestDenom = vecPrivateSendDenominations.back();
-        // Make outputs by looping through denominations, from large to small
-        for (const auto& nDenom : vecPrivateSendDenominations)
-        {
-            for (const auto& out : vCoins)
-            {
-                //make sure it's the denom we're looking for, round the amount up to smallest denom
-                if(out.tx->tx->vout[out.i].nValue == nDenom && nValueRet + nDenom < nTargetValue + nSmallestDenom) {
-                    COutPoint outpoint = COutPoint(out.tx->GetHash(),out.i);
-                    int nRounds = GetCappedOutpointPrivateSendRounds(outpoint);
-                    // make sure it's actually anonymized
-                    if(nRounds < privateSendClient.nPrivateSendRounds) continue;
-                    nValueRet += nDenom;
-                    setCoinsRet.insert(std::make_pair(out.tx, out.i));
+        std::vector<CAmount> vecDenominationsAsAFee = vecPrivateSendDenominations;
+        std::reverse(vecDenominationsAsAFee.begin(), vecDenominationsAsAFee.end());
+
+        // Try to fit into fee that matches one of denominations, from small to large
+        for (const auto& nDenomAsAFee : vecDenominationsAsAFee) {
+            CAmount nMaxPSFee = std::min(nDenomAsAFee, maxTxFee);
+            // Make outputs by looping through denominations, from large to small
+            for (const auto& nDenom : vecPrivateSendDenominations) {
+                for (const auto& out : vCoins) {
+                    // Make sure it's the denom we're looking for, round the amount up to current max fee
+                    if (out.tx->tx->vout[out.i].nValue == nDenom && nValueRet + nDenom < nTargetValue + nMaxPSFee) {
+                        COutPoint outpoint = COutPoint(out.tx->GetHash(),out.i);
+                        int nRounds = GetRealOutpointPrivateSendRounds(outpoint);
+                        // Make sure it's actually anonymized
+                        if (nRounds < privateSendClient.nPrivateSendRounds) continue;
+                        nValueRet += nDenom;
+                        setCoinsRet.insert(std::make_pair(out.tx, out.i));
+                        if (nValueRet >= nTargetValue) return true; // Done, no need to look any further
```

Comment:
I'm worried that as we are adding inputs it doesn't appear we are increasing the `nTargetValue` nor accounting for the fees of the added inputs... Could this fail in a way that the amount of the denoms is enough to cover the amount being sent but not enough to cover the amount + the fee resulting in a tx failing?

Reply:
I was thinking about this and no, technically, this should be good. But then I realised that even in this case we are probably doing it wrong (it's not the job of `SelectCoins` to guess fees/pick the best set). So, it was a very good question actually :) Pls see #2371 for an alternative solution.